lib/os/hash_map_oa_lp.c

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (c) 2022 Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <zephyr/sys/hash_map.h>
+#include <zephyr/sys/hash_map_oa_lp.h>
+#include <zephyr/sys/util.h>
+
+enum bucket_state {
+	UNUSED,
+	USED,
+	TOMBSTONE,
+};
+
+struct oalp_entry {
+	uint64_t key;
+	uint64_t value;
+	enum bucket_state state;
+};
+
+static struct oalp_entry *sys_hashmap_oa_lp_find(const struct sys_hashmap *map, uint64_t key,
+						 bool used_ok, bool unused_ok, bool tombstone_ok)
+{
+	size_t i;
+	size_t j;
+	struct oalp_entry *entry = NULL;
+	const size_t n_buckets = map->data->n_buckets;
+	uint32_t hash = map->hash_func(&key, sizeof(key));
+	struct oalp_entry *const buckets = map->data->buckets;
+
+	if (buckets == NULL) {
+		return NULL;
+	}
+
+	__ASSERT_NO_MSG(n_buckets != 0);
+
+	for (i = 0, j = hash % n_buckets; i < n_buckets; ++i, ++j, j %= n_buckets) {
```

Comment:
Maybe we should be making sure that n_buckets is a power of two and using `&` instead of modulus to avoid divides? And modulus is not just one divide, it's multiple, right? And this is in the search path.

Probably not a serious consideration in this first version though, maybe better to just make sure it works.

Reply:
Would make sense. It would need to be enforced that the table size is always a PoT